### PR TITLE
[Peterborough] Speed up WasteWorks by calling Bartec API in parallel

### PR DESCRIPTION
**Pull request by davea**
WasteWorks was making 6 sequential calls to the Bartec API on every page load which made the process feel quite sluggish - often taking up to 3 or 4 seconds for each page.

This PR factors out the existing Echo parallel API code into a `Role` which the Bartec module then uses. The Peterborough cobrand takes advantage of this which also has the side effect of caching results in the user's session, making page loads effectively instant.

[skip changelog]

### CODE REVIEW

File: bin/fixmystreet.com/call-wasteworks-backend [hunk 1]
```diff
@@ -17,22 +17,26 @@ use Getopt::Long::Descriptive;
...
     '%c %o',
     ['cobrand=s', 'which cobrand configuration to use'],
+    ['backend=s', 'which backend type (echo/bartec)'],
     ['out=s', 'where to output data'],
     ['calls=s', 'JSON of what API calls to make'],
     ['help|h', "print usage message and exit" ],
 );
 $usage->die if $opts->help;
-$usage->die unless $opts->cobrand && $opts->out && $opts->calls;
+$usage->die unless $opts->cobrand && $opts->out && $opts->calls && $opts->backend;
```
Comment (dracos) [inline bin/fixmystreet.com/call-wasteworks-backend]: Have just pushed a change I kept meaning to PR and not getting round to, to use this at the command line more easily, good for debugging something.

Comment (davea) [reply bin/fixmystreet.com/call-wasteworks-backend]: Nice, change looks good to me.

**Review by nephila-nacrea — APPROVED** [review]
I'm approving because I can't spot any mistakes, but I do have some questions :-)

File: perllib/FixMyStreet/App/Controller/Waste.pm [hunk 2]
```diff
@@ -580,7 +580,7 @@ sub property : Chained('/') : PathPart('waste') : CaptureArgs(1) {
     $c->forward('/auth/get_csrf_token');
 
     # clear this every time they visit this page to stop stale content.
-    if ( $c->req->path =~ m#^waste/\d+$# ) {
+    if ( $c->req->path =~ m#^waste/[:\w ]+$#i ) {
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/App/Controller/Waste.pm:583]: This is for the main "your bin days" page at `/waste/<property id>`. Bromley/Kingston use numeric IDs, but Peterborough's are of the form `postcode:uprn` so the regex now copes with those.

File: perllib/FixMyStreet/Cobrand/Peterborough.pm [hunk 3]
```diff
@@ -553,13 +565,27 @@ sub bin_services_for_address {
...
-    my $jobs = $bartec->Jobs_Get($property->{uprn});
-    my $schedules = $bartec->Features_Schedules_Get($property->{uprn});
-    my $job_dates = relevant_jobs($bartec, $property->{uprn}, $schedules);
-    my $events_uprn = $bartec->Premises_Events_Get($property->{uprn});
-    my $events_usrn = $bartec->Streets_Events_Get($property->{usrn});
-    my $open_requests = $self->open_service_requests_for_uprn($property->{uprn}, $bartec);
+    my $uprn = $property->{uprn};
+
+    my @calls = (
+        Jobs_Get => [ $uprn ],
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/Cobrand/Peterborough.pm:571]: Apparently so! I tried without the `[]` and it didn't work.

[the remark this answers is not in the document]
Comment (dracos) [reply perllib/FixMyStreet/Cobrand/Peterborough.pm:571]: Yeah, `_parallel_api_calls` requires that the `$args` are provided in an arrayref.

File: perllib/FixMyStreet/Roles/ParallelAPI.pm [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        '--backend', $type,
+        '--out', $tmp,
+        '--calls', encode_json(\@_),
+    );
+    my $start = Time::HiRes::time();
+
+    # We cannot fork directly under mod_fcgid, so
+    # call an external script that calls back in.
+    my $data;
+    # uncoverable branch false
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/Roles/ParallelAPI.pm:32]: [It tells `Devel::Cover`](https://metacpan.org/pod/Devel::Cover#UNCOVERABLE-CRITERIA) that the false path of the `if` clause isn't covered by tests and shouldn't be counted in the coverage stats.

File: perllib/FixMyStreet/Roles/ParallelAPI.pm [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    if (FixMyStreet->test_mode || $self->sample_data) {
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/Roles/ParallelAPI.pm:33]: `_parallel_api_calls` calls `fork` to run the API requests in parallel, which we can't do when running under `mod_fcgid` (ie, when running in a production webserver). `fork` _can_ be called when running tests, so this skips the `system` call to `call-wasteworks-backend` and directly runs the code that that script would.

File: perllib/FixMyStreet/Cobrand/Peterborough.pm [hunk 6]
```diff
@@ -414,6 +414,18 @@ sub clear_cached_lookups_postcode {
...
 }
 
+sub clear_cached_lookups_property {
+    my ($self, $uprn) = @_;
+
+    my $key = "peterborough:bartec:look_up_property:$uprn";
+    delete $self->{c}->session->{$key};
+    $key = "peterborough:bartec:bin_services_for_address:$uprn";
+    delete $self->{c}->session->{$key};
+    $key = "peterborough:bartec:property_attributes:$uprn";
```
[the remark this answers is not in the document]
Comment (davea) [reply perllib/FixMyStreet/Cobrand/Peterborough.pm]: I've gone one better and put it in a `foreach` in 6e7b8ac58d2db01781e119408e0010a6b8352306 :)

**Pull request merged**